### PR TITLE
Implement python drop_table

### PR DESCRIPTION
**Pull request by szehon-ho**
Implemented drop_table functionality in Python (based on Java functionality)
--Open manifest files to get list of data files
--Drop metadata, snapshot, manifest, and data files

Testing: added basic unit test

Changed flake8 for non-related issue that prevent tests from running (apparently it means, too many if-else)
--iceberg/api/expressions/literals.py:39:5: C901 'Literals.from_' is too complex (12)
--iceberg/api/expressions/literals.py:81:5: C901 'Literal.of' is too complex (11)

### CODE REVIEW

**Comment by szehon-ho** [comment]
This is the PR for #2266

**Comment by rymurr** [comment]
thanks @szehon-ho! I should get time to review in the next day or two.

**Comment by szehon-ho** [comment]
Great,  no rush, thanks!

File: python/iceberg/hive/hive_tables.py
Comment (rymurr) [inline python/iceberg/hive/hive_tables.py:60]: `s.manifest_location` is a property, `s.manifest_location()` should throw an error wen called like that shouldn't it?

File: python/iceberg/hive/hive_tables.py
Comment (rymurr) [inline python/iceberg/hive/hive_tables.py:44]: can you add mypy annotations to any new code please?

File: python/iceberg/hive/hive_tables.py
Comment (rymurr) [inline python/iceberg/hive/hive_tables.py:53]: broadly speaking I prefer the use of iterators/itertools where possible. Currently I think the underlying data structures are all `list` structures and aren't lazy so the use of iterators to collect files for deletion is a bit distracting.

File: python/tests/hive/test_hive_tables.py
Comment (rymurr) [inline python/tests/hive/test_hive_tables.py:166]: I think it would be good to test purge as well tbh, most of the code complexity in this change lives there.

**Review by rymurr — CHANGES_REQUESTED** [review]
Thanks a lot @szehon-ho looks good! Just a few minor comments.

**Comment by szehon-ho** [comment]
Thanks @rymurr  for the quick review, i will take a look next week at make the changes!

File: python/iceberg/hive/hive_tables.py
Comment (TGooch44) [reply python/iceberg/hive/hive_tables.py:65]: @rymurr No good reason to not switch to using threading directly if it's more clear.  asyncio would probably be better, but I'll admit that I'm not that familiar with developing in the python async model.  I can probably spend sometime getting caught up on that.

**Comment by TGooch44** [comment]
I should have some time to review this over the weekend as well

File: python/iceberg/hive/hive_tables.py
Comment (rymurr) [reply python/iceberg/hive/hive_tables.py:65]: I think async is cooler but I don't know if adding async here is really going to do a lot of good. Complicates the code and reduce the number of people who can read it so maybe (for now) threading module is best. What do you guys think?

File: python/iceberg/hive/hive_tables.py
Comment (szehon-ho) [reply python/iceberg/hive/hive_tables.py:60]: Yes weird, will check

File: python/iceberg/hive/hive_tables.py
Comment (szehon-ho) [reply python/iceberg/hive/hive_tables.py:65]: OK let me know, to be honest, I'm not familiar with async either.

File: python/iceberg/hive/hive_tables.py
Comment (szehon-ho) [reply python/iceberg/hive/hive_tables.py:67]: Got it

File: python/iceberg/hive/hive_tables.py
Comment (szehon-ho) [reply python/iceberg/hive/hive_tables.py:68]: Can add a check in delete_file first

File: python/tox.ini
Comment (szehon-ho) [reply python/tox.ini:91]: Will give a try

File: python/tests/hive/test_hive_tables.py
Comment (szehon-ho) [reply python/tests/hive/test_hive_tables.py:166]: Will do

**Review by szehon-ho — COMMENTED** [review]
Had a chance to look through the review comments, just wanted to clarify one thing about the generators, will take a look at the rest.

File: python/tox.ini
Comment (szehon-ho) [reply python/tox.ini]: It seems it was not on the exact line, at least when I ran tox?  I moved it to the line that fixed it for me.

File: python/iceberg/hive/hive_tables.py
Comment (szehon-ho) [reply python/iceberg/hive/hive_tables.py]: Done

File: python/iceberg/hive/hive_tables.py
Comment (szehon-ho) [reply python/iceberg/hive/hive_tables.py]: Yea it took me a bit to find the mystery.  In the implementation of the property, manifest_list does  "FileSystemInputFile.location" which is itself a method.  Changed the implementation of the property to fix this..

File: python/iceberg/hive/hive_tables.py
Comment (szehon-ho) [reply python/iceberg/hive/hive_tables.py]: I tried for awhile, I hit some existing issues like:
`  File "/iceberg/python/iceberg/core/base_snapshot.py", line 131, in get_filtered_manifest
    reader = ManifestReader.read(self.ops.new_input_file(path))
AttributeError: 'BaseSnapshot' object has no attribute 'ops'
`
I could fix those, but was wondering if it's not better to just have these two lines here as well, if it's pretty static logic (save the reference to self.conf)?  DataTableScan does the same thing.  The problem with calling the snapshot version  is that then I'd have to pass the snapshot along with the manifest in my iterator stuff and it makes it slightly even less readable.

File: python/iceberg/hive/hive_tables.py
Comment (szehon-ho) [reply python/iceberg/hive/hive_tables.py]: Done

File: python/iceberg/hive/hive_tables.py
Comment (szehon-ho) [reply python/iceberg/hive/hive_tables.py]: Added a deduplication for the iterator

File: python/iceberg/hive/hive_tables.py
Comment (szehon-ho) [reply python/iceberg/hive/hive_tables.py]: Done

File: python/tests/hive/test_hive_tables.py
Comment (szehon-ho) [reply python/tests/hive/test_hive_tables.py]: Added a test with a bunch of mocks to test that part.

File: python/iceberg/hive/hive_tables.py
Comment (szehon-ho) [reply python/iceberg/hive/hive_tables.py]: Spent some time to find the mystery.  The property getter was calling "FileSystemInputFile.location" which is a method.  Fixed it on the property itself.

File: python/iceberg/hive/hive_tables.py
Comment (szehon-ho) [reply python/iceberg/hive/hive_tables.py]: I tried a bit for this.  First I hit an error:

  File "/iceberg/python/iceberg/core/base_snapshot.py", line 131, in get_filtered_manifest
    reader = ManifestReader.read(self.ops.new_input_file(path))
AttributeError: 'BaseSnapshot' object has no attribute 'ops'

I could fix the error but wondering if it's worth it over two lines?  DataTableScan also has these two lines.  This logic is mostly static , save for self.conf reference.  Otherwise I'd have to pass the snapshot around along with the manifest in the iterators, which makes the code even harder to read.

File: python/iceberg/hive/hive_tables.py
Comment (szehon-ho) [reply python/iceberg/hive/hive_tables.py]: Done

File: python/iceberg/hive/hive_tables.py
Comment (szehon-ho) [reply python/iceberg/hive/hive_tables.py]: Added a de-duplication in the iterator.

File: python/iceberg/hive/hive_tables.py
Comment (szehon-ho) [reply python/iceberg/hive/hive_tables.py]: Done

File: python/tests/hive/test_hive_tables.py
Comment (szehon-ho) [reply python/tests/hive/test_hive_tables.py]: Spent some time to add a second test, with mocking all the method calls to test the new code.

File: python/iceberg/hive/hive_tables.py
Comment (szehon-ho) [reply python/iceberg/hive/hive_tables.py]: Done

**Comment by szehon-ho** [comment]
@rymurr , I made some of the suggested code changes and replied where I could not, thanks again for reviewing it

File: python/tox.ini
Comment (szehon-ho) [reply python/tox.ini]: It still fails when running tox, at least for me.  I moved the annotations to the line that fixed it for me.

File: python/iceberg/hive/hive_tables.py
Comment (szehon-ho) [reply python/iceberg/hive/hive_tables.py]: By the way the TableOps implementation was passing too many arguments, fixed it.

File: python/iceberg/api/manifest_file.py
Comment (rymurr) [inline python/iceberg/api/manifest_file.py:30]: nit: unrelated whitespace change

File: python/tox.ini
Comment (rymurr) [reply python/tox.ini]: it depends on the python you ran it on. py3.8 changed the parse rules and the most recent linters fix it. #2268 changed the linters to run on py38 so should work fine on master now

File: python/iceberg/hive/hive_tables.py
Comment (rymurr) [reply python/iceberg/hive/hive_tables.py]: cool, threading it is :+1:

File: python/iceberg/hive/hive_tables.py
Comment (rymurr) [inline python/iceberg/hive/hive_tables.py:95]: I think it might be easier to just catch an exceptoion in `_delete_file`

File: python/iceberg/hive/hive_tables.py
Comment (rymurr) [inline python/iceberg/hive/hive_tables.py:54]: I had something a bit simpler in mind. I think we can avoid some of the itertools stuff by just deleting as we find files. This also implies that the `_delete_file` method catches deletes that happen twice

``` python
            if metadata is not None:
                with Pool(self.conf.get(WORKER_THREAD_POOL_SIZE_PROP,
                          cpu_count())) as delete_pool:
                    manifest_lists_to_delete = [ops.current_metadata_location]

                    for s in metadata.snapshots:
                        delete_pool.map(self._delete_file(ops), (i.path() for i in s.get_filtered_manifests().iterator()))
                        delete_pool.map(self._delete_file(ops), s.manifests)
                        if s.manifest_location is not None:
                            manifest_lists_to_delete.append(s.manifest_location)
                    delete_pool.map(self._delete_file(ops), manifest_lists_to_delete)
```

I didn't test this code so it may not be quite right. And some manifests mentioned in later snapshots may already be deleted so can't be read (but can then just be skipped).

File: python/tests/hive/test_hive_tables.py
Comment (rymurr) [inline python/tests/hive/test_hive_tables.py:192]: I think these could be useful in `test_helpers.py`

**Review by rymurr — COMMENTED** [review]
Looking good! Thanks again @szehon-ho !

File: python/iceberg/hive/hive_tables.py
Comment (szehon-ho) [reply python/iceberg/hive/hive_tables.py]: Done

**Comment by szehon-ho** [comment]
@rymurr Thanks a lot Ryan for helping me with this change.  Did the suggestions, when you have another chance

File: python/iceberg/core/base_snapshot.py
Comment (rymurr) [inline python/iceberg/core/base_snapshot.py:129]: how come the `None` were removed?

File: python/iceberg/hive/hive_tables.py
Comment (rymurr) [inline python/iceberg/hive/hive_tables.py:42]: can you add mypy annotations to the arguments here

File: python/iceberg/hive/hive_tables.py
Comment (rymurr) [inline python/iceberg/hive/hive_tables.py:71]: mypy here too pls

**Review by rymurr — COMMENTED** [review]
looking good @szehon-ho just a few minor comments left :-)

File: python/iceberg/core/base_snapshot.py
Comment (szehon-ho) [reply python/iceberg/core/base_snapshot.py:129]: Hi , I didn't remove but rather added None to be the default argument for (part_filter, row_filter, column)

They actually are not used in method itself.. but not sure if anyone was calling with those them.  I can just remove them as well?

Comment (rymurr) [reply python/iceberg/core/base_snapshot.py:129]: Doh! Misread the diff. I like your way.

File: python/iceberg/hive/hive_tables.py
Comment (szehon-ho) [reply python/iceberg/hive/hive_tables.py]: Done

File: python/iceberg/hive/hive_tables.py
Comment (szehon-ho) [reply python/iceberg/hive/hive_tables.py]: Done

**Comment by szehon-ho** [comment]
Done, thanks @rymurr !  To be honest, its my first python change in quite awhile :)

**Comment by rymurr** [comment]
looks great @szehon-ho I will approve once the tests pass. Thanks a lot!

**Comment by szehon-ho** [comment]
OK let me fix the linting

**Comment by szehon-ho** [comment]
Finally all green!

**Review by rymurr — APPROVED** [review]
LGTM, thanks @szehon-ho !

**Comment by szehon-ho** [comment]
Thanks again @rymurr  for your time!

**Comment by szehon-ho** [comment]
@rdblue do you think we can merge and /or take a look?  Thanks

**Comment by TGooch44** [comment]
Ryan(Blue) is out on family leave right now, maybe @danielcweeks can take a look at merging.

**Comment by szehon-ho** [comment]
Thanks for letting me know, @TGooch44

**Comment by rymurr** [comment]
Hey @danielcweeks did you want to give this a review or can I go ahead and merge?

File: python/iceberg/hive/hive_tables.py
Comment (danielcweeks) [inline python/iceberg/hive/hive_tables.py:78]: Feels like we should at least log warning level message in the event the files cannot be deleted.  It looks like we're silently swallowing files that would be dangling after this operation.

**Comment by danielcweeks** [comment]
@szehon-ho Just one comment about swallowing failures, bot other than that, it looks good to me.  @rymurr feel free to merge when you think it's ready.

File: python/iceberg/hive/hive_tables.py
Comment (rymurr) [reply python/iceberg/hive/hive_tables.py:78]: agreed, perhaps w/ a `set` to prevent multiple alerts?

Comment (szehon-ho) [reply python/iceberg/hive/hive_tables.py:78]: Yea sure guys.   @rymurr I suppose we need to protect the set with lock to be thread-safe?  Is it worth it (I imagine it slows down a bit)?

**Comment by szehon-ho** [comment]
@rymurr  thanks for remembering, i was going to ping you as well not that you are committer :).  I updated with a basic log, just wondering about the set, let me know and i can change as well.

**Comment by szehon-ho** [comment]
Whoops, now linter is good.

File: python/iceberg/hive/hive_tables.py
Comment (rymurr) [inline python/iceberg/hive/hive_tables.py:73]: I think it might be nicer to call `self._delete_file` once in the above method and add a `set` here. Then log only files that aren't in the set yet.

File: python/iceberg/hive/hive_tables.py
Comment (rymurr) [inline python/iceberg/hive/hive_tables.py:79]: nit: this `pass` is redundant now

File: python/iceberg/hive/hive_tables.py
Comment (szehon-ho) [reply python/iceberg/hive/hive_tables.py:73]: Hm sorry can you elaborate?

Comment (szehon-ho) [reply python/iceberg/hive/hive_tables.py:73]: Ah got it,  thanks.  Should we just skip deletion itself if its in seen?  And in python, need to protect this set with a lock.

Comment (rymurr) [reply python/iceberg/hive/hive_tables.py:73]: > Ah got it, thanks. Should we just skip deletion itself if its in seen?

even better :-)

**Comment by szehon-ho** [comment]
@rymurr implemented the set, also add a set lock to protect.  I split _delete_files into a separate class as linter was complaining that the method was too complex.

The alternate is to have a filter method like `delete_pool.map(delete_files, filter(...))` , but I guess it's a bit similar to what I had much earlier, so I wrote it in this new way.

**Comment by szehon-ho** [comment]
Also forgot, I added new test data with replicated metadata-files and checks to validate this in test_drop_tables.

File: python/tests/hive/test_hive_tables.py
Comment (szehon-ho) [inline python/tests/hive/test_hive_tables.py:222]: Validation for files referred to multiple times by various snapshots

**Review by rymurr — APPROVED** [review]
Looks great @szehon-ho, thanks for your patience!

**Comment by szehon-ho** [comment]
No problem, thanks @rymurr for your time and great code suggestions as well

**Pull request merged**
